src/workerd/api/performance.h

```diff
@@ -0,0 +1,420 @@
+
+    JSG_STRUCT(detail, startTime);
+  };
+
+  jsg::JsValue getDetail(jsg::Lock& js) {
+    KJ_IF_SOME(value, detail) {
+      return value.getHandle(js);
+    }
+    return js.null();
+  }
+
+  JSG_RESOURCE_TYPE(PerformanceMark) {
+    JSG_INHERIT(PerformanceEntry);
+    JSG_READONLY_PROTOTYPE_PROPERTY(detail, getDetail);
+  }
+
+ private:
+  jsg::Optional<jsg::JsRef<jsg::JsValue>> detail;
+};
+
+class PerformanceMeasure: public PerformanceEntry {
+ public:
+  PerformanceMeasure(kj::String name, uint32_t startTime, uint32_t duration)
+      : PerformanceEntry(kj::mv(name), kj::str("measure"), startTime, duration) {}
+
+  struct Entry {
+    kj::String entryType;
+    kj::String name;
+    kj::OneOf<kj::Date, uint32_t> startTime;
+    uint32_t duration;
+    jsg::Optional<jsg::JsRef<jsg::JsObject>> detail;
+
+    JSG_STRUCT(entryType, name, startTime, duration, detail);
+  };
+
+  struct Options {
+    jsg::Optional<jsg::JsRef<jsg::JsObject>> detail;
+    jsg::Optional<uint32_t> start;
+    jsg::Optional<uint32_t> duration;
+    jsg::Optional<uint32_t> end;
+
+    JSG_STRUCT(detail, start, duration, end);
+  };
+
+  jsg::JsValue getDetail(jsg::Lock& js) {
+    KJ_IF_SOME(value, detail) {
+      return value.getHandle(js);
+    }
+    return js.null();
+  }
+
+  JSG_RESOURCE_TYPE(PerformanceMeasure) {
+    JSG_INHERIT(PerformanceEntry);
+    JSG_READONLY_PROTOTYPE_PROPERTY(detail, getDetail);
+  }
+
+ private:
+  jsg::Optional<jsg::JsRef<jsg::JsValue>> detail;
+};
+
+class PerformanceResourceTiming: public PerformanceEntry {
+ public:
+  PerformanceResourceTiming(kj::String name, uint32_t startTime, uint32_t duration)
+      : PerformanceEntry(kj::mv(name), kj::str("resource"), startTime, duration) {}
+
+  uint32_t getConnectEnd() {
+    return connectEnd;
+  }
+  uint32_t getConnectStart() {
+    return connectStart;
+  }
+  uint32_t getDecodedBodySize() {
+    return decodedBodySize;
+  }
+  uint32_t getDomainLookupEnd() {
+    return domainLookupEnd;
+  }
+  uint32_t getDomainLookupStart() {
+    return domainLookupStart;
+  }
+  uint32_t getEncodedBodySize() {
+    return encodedBodySize;
+  }
+  uint32_t getFetchStart() {
+    return fetchStart;
+  }
+  jsg::JsString getInitiatorType(jsg::Lock& js) {
+    return js.str(initiatorType);
+  }
+  jsg::JsString getNextHopProtocol(jsg::Lock& js) {
+    return js.str(nextHopProtocol);
+  }
+  uint32_t getRedirectEnd() {
+    return redirectEnd;
+  }
+  uint32_t getRedirectStart() {
+    return redirectStart;
+  }
+  uint32_t getRequestStart() {
+    return requestStart;
+  }
+  uint32_t getResponseEnd() {
+    return responseEnd;
+  }
+  uint32_t getResponseStart() {
+    return responseStart;
+  }
+  uint32_t getResponseStatus() {
+    return responseStatus;
+  }
+  jsg::Optional<uint32_t> getSecureConnectionStart() {
+    return secureConnectionStart;
+  }
+  uint32_t getTransferSize() {
+    return transferSize;
+  }
+  uint32_t getWorkerStart() {
+    return workerStart;
+  }
```

Comment:
better to just return literals since these are all non-functional anyway

```suggestion
  uint32_t getConnectEnd() {
    return 0;
  }
  uint32_t getConnectStart() {
    return 0;
  }
  uint32_t getDecodedBodySize() {
    return 0;
  }
  uint32_t getDomainLookupEnd() {
    return 0;
  }
  uint32_t getDomainLookupStart() {
    return 0;
  }
  uint32_t getEncodedBodySize() {
    return 0;
  }
  uint32_t getFetchStart() {
    return 0;
  }
  jsg::JsString getInitiatorType(jsg::Lock& js) {
    return js.str();
  }
  jsg::JsString getNextHopProtocol(jsg::Lock& js) {
    return js.str();
  }
  uint32_t getRedirectEnd() {
    return 0;
  }
  uint32_t getRedirectStart() {
    return 0;
  }
  uint32_t getRequestStart() {
    return 0;
  }
  uint32_t getResponseEnd() {
    return 0;
  }
  uint32_t getResponseStart() {
    return 0;
  }
  uint32_t getResponseStatus() {
    return 0;
  }
  jsg::Optional<uint32_t> getSecureConnectionStart() {
    return kj::none;
  }
  uint32_t getTransferSize() {
    return 0;
  }
  uint32_t getWorkerStart() {
    return 0;
  }
```